File: tools/_build_relution_import_artifacts_modules/recommendation_implementation.py
"""Implementation-category projection for normalized recommendations."""

from typing import Any

from recommendation_mapping import unique_preserving_order

from .mapping_helpers import exact_mappings
from .recommendation_implementation_category import implementation_category


def implementation_for(source: str, recommendation: dict[str, Any], fallback_translations: list[dict[str, Any]]) -> dict[str, Any]:
    """Build implementation category, surfaces, and import path metadata."""
    relution_mapping = recommendation.get("relutionMapping", {})
    exact = exact_mappings(recommendation)
    candidate_surfaces = unique_preserving_order(candidate.get("kind") for candidate in relution_mapping.get("candidates", []) if isinstance(candidate, dict) and isinstance(candidate.get("kind"), str))
    category, blocking_reasons = implementation_category({"source": source, "recommendation": recommendation, "exact": exact, "candidateSurfaces": candidate_surfaces, "fallbackTranslations": fallback_translations, "notes": [str(note) for note in relution_mapping.get("notes", []) if isinstance(note, str) and note]})
    return {"category": category, "surfaces": implementation_surfaces(exact, candidate_surfaces, fallback_translations), "importableVia": implementation_import_paths(exact), "blockingReasons": blocking_reasons}


def implementation_surfaces(exact: list[dict[str, Any]], candidate_surfaces: list[str], fallback_translations: list[dict[str, Any]]) -> list[str]:
    """Return stable implementation-surface labels."""
    return unique_preserving_order([*(mapping["kind"] for mapping in exact), *candidate_surfaces, *( ["helper"] if fallback_translations else [])])


def implementation_import_paths(exact: list[dict[str, Any]]) -> list[str]:
    """Return supported import paths for exact mappings."""
    return unique_preserving_order([*( ["ruleset-import"] if exact else []), *( ["apply-json"] if any(mapping.get("kind") == "relution-native" for mapping in exact) else [])])

File: web/src/editor/recommendation-implementation.ts
// Supports recommendation editor views, filtering, and display metadata.
import type { RecommendationImplementation, RecommendationRecord } from "../../../src/recommendation-types.js";
import { uniqueStrings } from "../../../src/utils/json-guards.js";

export function implementationOf(recommendation: RecommendationRecord): RecommendationImplementation {
  if (recommendation.implementation !== undefined) {
    return recommendation.implementation;
  }
  const fallbackTranslations = recommendation.fallbackTranslations ?? [];
  const surfaces = uniqueStrings([
    ...recommendation.relutionMapping.candidates.map((candidate) => candidate.kind),
    ...recommendation.relutionMapping.rulesetMappings.map((mapping) => mapping.kind),
    ...(fallbackTranslations.length > 0 ? ["helper"] : []),
  ], { sort: true }) as RecommendationImplementation["surfaces"];
  return inferredImplementation(recommendation, surfaces, fallbackTranslations.length > 0);
}

function inferredImplementation(
  recommendation: RecommendationRecord,
  surfaces: RecommendationImplementation["surfaces"],
  hasFallbackTranslations: boolean,
): RecommendationImplementation {
  const { relutionMapping } = recommendation;
  if (relutionMapping.status === "exact") {
    return {
      category: "relution-achievable",
      surfaces,
      importableVia: relutionMapping.rulesetMappings.some((mapping) => mapping.kind === "relution-native")
        ? ["apply-json", "ruleset-import"]
        : ["ruleset-import"],
      blockingReasons: relutionMapping.notes,
    };
  }
  return nonExactImplementation(recommendation, surfaces, hasFallbackTranslations);
}

function nonExactImplementation(
  recommendation: RecommendationRecord,
  surfaces: RecommendationImplementation["surfaces"],
  hasFallbackTranslations: boolean,
): RecommendationImplementation {
  const category = recommendation.relutionMapping.candidates.length > 0
    ? "relution-partial"
    : hasFallbackTranslations
      ? "helper-only"
      : "gap";
  return { category, surfaces, importableVia: [], blockingReasons: recommendation.relutionMapping.notes };
}

export function importabilityLabel(implementation: RecommendationImplementation): string {
  return implementation.importableVia.length === 0 ? "Info only" : `Importable via ${implementation.importableVia.join(", ")}`;
}

export function categoryLabel(category: string): string {
  return ({
    "relution-achievable": "Achievable",
    "relution-partial": "Partial",
    "helper-only": "Helper only",
    gap: "Gap",
  } as Record<string, string>)[category] ?? category;
}

export function surfaceLabel(surface: string): string {
  return ({
    "relution-native": "Native",
    "apple-mobileconfig": "Apple mobileconfig",
    "apple-schema-profile": "Apple schema",
    helper: "Helper",
  } as Record<string, string>)[surface] ?? surface;
}
